go/src/koding/socialapi/transport_test.go

```diff
@@ -0,0 +1,142 @@
+package socialapi_test
+
+import (
+	"encoding/json"
+	"log"
+	"net"
+	"net/http"
+	"strings"
+	"testing"
+
+	"koding/kites/tunnelproxy/discover/discovertest"
+	"koding/socialapi"
+)
+
+func TestTransport(t *testing.T) {
+	var storage TrxStorage
+	var auth = NewFakeAuth()
+
+	s, err := discovertest.NewServer(http.HandlerFunc(auth.GetSession))
+	if err != nil {
+		t.Fatalf("NewServer()=%s", err)
+	}
+	defer s.Close()
+
+	cache := socialapi.NewCache(auth.Auth)
+	cache.Storage = &storage
+
+	users := []*socialapi.Session{
+		{Username: "user1", Team: "foobar"},
+		{Username: "user2", Team: "foobar"},
+		{Username: "user3", Team: "foobar"},
+		{Username: "user", Team: "team"},
+	}
+
+	cases := []struct {
+		name  string
+		sess  *socialapi.Session // client
+		errs  []error            // fake errors for endpoint
+		codes []int              // fake response codes for endpoint
+		err   error              // final error from client
+		trx   TrxStorage         // underlying cache operations
+	}{{
+		"new user1",
+		users[0],
+		nil, nil, nil,
+		TrxStorage{
+			{Type: "get", Session: users[0]},
+			{Type: "set", Session: users[0]},
+		},
+	}, {
+		"cached user1",
+		users[0],
+		nil, nil, nil,
+		TrxStorage{
+			{Type: "get", Session: users[0]},
+		},
+	}, {
+		"new user2",
+		users[1],
+		nil, nil, nil,
+		TrxStorage{
+			{Type: "get", Session: users[1]},
+			{Type: "set", Session: users[1]},
+		},
+	}, {
+		"new user3 with an error",
+		users[2],
+		[]error{&net.DNSError{IsTemporary: true}}, nil, nil,
+		TrxStorage{
+			{Type: "get", Session: users[2]},
+			{Type: "set", Session: users[2]},
+		},
+	}, {
+		"cached user3 with an error and response codes",
+		users[2],
+		[]error{&net.DNSError{IsTemporary: true}}, []int{401, 401}, nil,
+		TrxStorage{
+			{Type: "get", Session: users[2]},
+		},
+	}}
+
+	rec := &AuthRecorder{
+		AuthFunc: cache.Auth,
+	}
+
+	client := http.Client{
+		Transport: &socialapi.Transport{
+			RoundTripper: &FakeTransport{},
```

Comment:
> they are not faking or mocking any functionality.

which does not mean you cant, you can think them as small adapters and do what you want to do

for the hook debate (since you said it makes it worse), i find easier to reason about the cases/logic since they are in the code itself, not separated into another layer. i dont support having many hooks in the code itself too but i am also against having a big function and to be able to test adding another level of complexity with trying to fake the logical parts of it, which incidentally may lead to bugs in the tests itself(another question comes as how/should we (are going to) test the tests?). I would probably go with unit tests and smaller functions instead of having bigger components and mocks. 

in the end everyone has their own taste and there is no universal truth, at least i know of.